### PR TITLE
feat(post): interlinking

### LINKED ISSUE #293: New Feature - Interlinking Capability in Helix
SEO team has a request to add the ability to add theme relevant click links to blog article content. 

- SEO team would like to manage the keyword and link content through a spreadsheet of some sorts where they can regularly add, update keywords  
- System should support interlink updates to all new blog articles created on the Helix platform. The links can be added during publishing process (instead of at every update to minimize how many times we run a script to add links)
- System should support application of links on all migrated content from theblog.adobe.com, CMO.adobe.com and all the other blogs sites. 

@rofe @davidnuescheler - this is based on the email discussion we had on july 16 and july 17th.

### PR DESCRIPTION
**Pull request by rofe**
Fix #293 (replaces #421)

TODO:
- [x] remove debug console output

### CODE REVIEW

**Comment by rofe** [comment]
test with posts under https://interlink--theblog--adobe.hlx.page/

File: scripts/post.js [hunk 1]
```diff
@@ -332,6 +332,119 @@ function decorateImages() {
...
+ * intersection, the more specific (i.e. longer) match wins.
+ * @param {array} matches The existing matches
+ * @param {object} contender The match to check and add
+ */
+function checkAndAddMatch(matches, contender) {
+  const collisions = matches
+    // check for intersections
+    .filter((match) => {
+      if ( contender.start > match.end ||
+        (contender.start < match.start && contender.end < match.end)) {
```
Comment (tripodsan) [inline scripts/post.js:347]: isn't it enough just to check `contender.end < match.start || contender.start > match.end` ?

| case                | a.end < b.start | a.start > b.end |  ok |
|---------------------|-----------------|-----------------|-----|
| `[a---a] [b---b]`   | true            | false           | yes |  
| `[b---b] [a---a]`   | false           | true            | yes |
| `[a---[b---b]---a]` | false           | false           | no  |
| `[b---[a---a]---b]` | false           | false           | no  |
| `[a---[b---a]---b]` | false           | false           | no  |
| `[b---[a---b]---a]` | false           | false           | no  |

**Review by kptdobe — APPROVED** [review]

File: scripts/post.js [hunk 2]
```diff
@@ -332,6 +332,105 @@ function decorateImages() {
...
+      if (contender.end < match.start || contender.start > match.end) {
+        // no intersection with existing match
+        return false;
+      }
+      // contender starts or ends within existing match
+      return true;
+    })
+    // check for specificity
+    .filter((match) => {
+      if (contender.item.Keyword.length > match.item.Keyword.length) {
```
Comment (tripodsan) [inline scripts/post.js:355]: shouldn't the test for better match be before the overlap rejection?

otherwise you have:

text: 'The quick brown fox jumps over the lazy dog.'
kw1: 'fox'
text: 'The quick brown **fox** jumps over the lazy dog.'
kw2: 'brown fox' -> overlaps with kw1 -> rejected

I think if you sort the keywords by length descending, you don't need this check.

**Review by tripodsan — CHANGES_REQUESTED** [review]
I think there is a bug, when a shorter keyword is matched with an overlapping longer keyword.

File: scripts/post.js (hunk 2 above)
Comment (tripodsan) [reply scripts/post.js:355]: (a unit test, would allow to quickly test all edge cases :-)

Comment (rofe) [reply scripts/post.js:355]: I couldn't reproduce the `fox` keyword winning over the `brown fox`, but you were right with sorting keywords by length before matching: this produces better results with less code 🙂

**Comment by rofe** [comment]
@kptdobe please merge once it gets @tripodsan approval

File: scripts/post.js [hunk 3]
```diff
@@ -332,6 +332,74 @@ function decorateImages() {
...
+async function addInterLinks() {
+  const response = await fetch('/en/keywords.json');
+  if (response.ok) { 
+    const json = await response.json();
+    const keywords = (Array.isArray(json) ? json : json.data)
+      // scan article to filter keywords down to relevant ones
+      .filter(({ Keyword }) => document.querySelector('main').textContent.toLowerCase().indexOf(Keyword) !== -1)
+      // sort matches by length descending
+      .sort((a, b) => {
+        return b.length - a.length;
```
Comment (tripodsan) [inline scripts/post.js:348]:
```suggestion
        return b.Keyword.length - a.Keyword.length;
```
???

**Review by tripodsan — CHANGES_REQUESTED** [review]
why did you remove the overlap test ?

**Comment by rofe** [comment]
> why did you remove the overlap check ?

I thought you said I don't need it if I sort keywords by length descending 🙂 
But you're right, I need to add it back.

> I think the sort by length is wrong. it needs to compare Keyword.length

Oops, thanks for catching!

> add unit test!

We don't have a unit test setup for the project yet, only [smoke tests](https://github.com/adobe/theblog-tests/tree/master/test/smoke). That would be a bigger effort. I'm using manual testing for now.

**Comment by rofe** [comment]
Keywords are now sorted by length descending, so the above example results in the folloiwng behavior:
text: 'The quick brown fox jumps over the lazy dog.'
kw1: 'brown fox'
text: 'The quick ~~brown fox~~ jumps over the lazy dog.'
kw2: 'fox' -> overlaps with kw1 -> rejected

**Review by tripodsan — APPROVED** [review]
